include/rmm/mr/device/detail/arena.hpp

```diff
@@ -0,0 +1,415 @@
+  {
+    lock_guard lock(mtx_);
+    bool found = do_deallocate(p, bytes);
+    shrink_arena(stream);
+    return found;
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by `p`, keeping all free superblocks.
+   *
+   * This is done when deallocating across arenas. Since we don't have access to the CUDA stream
+   * associated with this arena, it's not safe to return superblocks.
+   *
+   * @param p Pointer to be deallocated
+   * @param bytes The size in bytes of the allocation. This must be equal to the
+   * value of `bytes` that was passed to the `allocate` call that returned `p`.
+   * @return true if the allocation is found, false otherwise
+   */
+  bool deallocate(void* p, std::size_t bytes)
+  {
+    lock_guard lock(mtx_);
+    return do_deallocate(p, bytes);
+  }
+
+  /**
+   * @brief Deallocate memory pointed to by `p`.
+   *
+   * @param p Pointer to be deallocated
+   * @param bytes The size in bytes of the allocation. This must be equal to the
+   * value of `bytes` that was passed to the `allocate` call that returned `p`.
+   * @return bool if the allocation is found
+   */
+  bool do_deallocate(void* p, std::size_t bytes)
+  {
+    auto b = free_block(p, bytes);
+    coalesce_block(b);
+    return b.is_valid();
+  }
+
+  /**
+   * @brief Does an arena handle blocks of this size?
+   *
+   * @param size The size in bytes to check
+   * @return true if blocks of this size are handled by an arena
+   */
+  static bool handles_size(std::size_t size) { return size <= maximum_allocation_size; }
+
+ private:
+  /// The maximum allocation size handled by arenas.
+  static constexpr std::size_t maximum_allocation_size = block::superblock_size / 2;
+
+  using lock_guard = std::lock_guard<std::mutex>;
+
+  /**
+   * @brief Get an available memory block of at least `size` bytes.
+   *
+   * @param size The number of bytes to allocate
+   * @return block A block of memory of at least `size` bytes
+   */
+  block get_block(std::size_t size)
+  {
+    // Find the first-fit free block.
+    block b = first_fit(size);
+    if (b.is_valid()) return b;
+
+    // No free blocks available, use the first free superblock.
+    b = first_superblock();
+    if (b.is_valid()) {
+      RMM_LOGGING_ASSERT(b.fits(size));
+      return b;
+    }
+
+    // No existing larger blocks available, so grow the arena and obtain a superblock.
+    return expand_arena();
+  }
+
+  /**
+   * @brief Get the first free block of at least `size` bytes.
+   *
+   * @param size The number of bytes to allocate
+   * @return block A block of memory of at least `size` bytes
+   */
+  block first_fit(std::size_t size)
+  {
+    auto const iter = std::find_if(
+      free_blocks_.cbegin(), free_blocks_.cend(), [size](block const& b) { return b.fits(size); });
+
+    if (iter == free_blocks_.cend()) {
+      return {};
+    } else {
+      // Remove the block from the free_list and return it.
+      block const found = *iter;
+      RMM_LOGGING_ASSERT(found.fits(size));
```

Comment:
Is this ASSERT needed? You just checked that it fits in the `find_if`. If it now doesn't fit, that would be very strange, and I'm not sure a logging assert here is going to help.